### PR TITLE
rafthttp: report unreachable status of the peer

### PR DESCRIPTION
**Pull request by yichengq**
for #2381

### CODE REVIEW

**Comment by xiang90** [comment]
have you tested this manually?

**Comment by yichengq** [comment]
Yes. I start node1 and node2 of a 3-member (hope to see unreachable), and add node3 later (hope to see reachable):
16:50:30 etcd2 | 2015/03/01 16:50:30 raft: 6e3bd23ae5f1eae0 became leader at term 9
16:50:30 etcd2 | 2015/03/01 16:50:30 raft.node: 6e3bd23ae5f1eae0 elected leader 6e3bd23ae5f1eae0 at term 9
16:50:30 etcd1 | 2015/03/01 16:50:30 raft.node: a8266ecf031671f3 elected leader 6e3bd23ae5f1eae0 at term 9
*16:50:30 etcd2 | 2015/03/01 16:50:30 raft: 6e3bd23ae5f1eae0 failed to send message to 924e2e83e93f2560 and changed it to be unreachable [next = 15232, match = 0, wait = 1]*
16:50:30 etcd1 | 2015/03/01 16:50:30 raft: a8266ecf031671f3 [logterm: 0, index: 15234] rejected msgApp [logterm: 9, index: 15234] from 6e3bd23ae5f1eae0
16:50:30 etcd2 | 2015/03/01 16:50:30 raft: 6e3bd23ae5f1eae0 received msgApp rejection(lastindex: 15232) from a8266ecf031671f3 for index 15234
16:50:30 etcd2 | 2015/03/01 16:50:30 raft: 6e3bd23ae5f1eae0 decreased progress of a8266ecf031671f3 to [next = 15233, match = 15232, wait = 0]
16:50:30 etcd2 | 2015/03/01 16:50:30 etcdserver: published {Name:infra2 ClientURLs:[http://localhost:4002]} to cluster e9c7614f68f35fb2
16:50:30 etcd1 | 2015/03/01 16:50:30 etcdserver: published {Name:infra1 ClientURLs:[http://localhost:4001]} to cluster e9c7614f68f35fb2
16:50:30 etcd2 | 2015/03/01 16:50:30 rafthttp: roundtripping error: error roundtripping to http://localhost:7003/raft/stream/6e3bd23ae5f1eae0: dial tcp 127.0.0.1:7003: connection refused
16:50:30 etcd2 | 2015/03/01 16:50:30 rafthttp: roundtripping error: error roundtripping to http://localhost:7003/raft/stream/message/6e3bd23ae5f1eae0: dial tcp 127.0.0.1:7003: connection refused
16:50:30 etcd1 | 2015/03/01 16:50:30 rafthttp: roundtripping error: error roundtripping to http://localhost:7003/raft/stream/a8266ecf031671f3: dial tcp 127.0.0.1:7003: connection refused
16:50:30 etcd1 | 2015/03/01 16:50:30 rafthttp: roundtripping error: error roundtripping to http://localhost:7003/raft/stream/message/a8266ecf031671f3: dial tcp 127.0.0.1:7003: connection refused
16:50:30 etcd2 | 2015/03/01 16:50:30 rafthttp: roundtripping error: error roundtripping to http://localhost:7003/raft/stream/message/6e3bd23ae5f1eae0: dial tcp 127.0.0.1:7003: connection refused
16:50:30 etcd2 | 2015/03/01 16:50:30 rafthttp: roundtripping error: error roundtripping to http://localhost:7003/raft/stream/6e3bd23ae5f1eae0: dial tcp 127.0.0.1:7003: connection refused
16:50:30 etcd1 | 2015/03/01 16:50:30 rafthttp: roundtripping error: error roundtripping to http://localhost:7003/raft/stream/message/a8266ecf031671f3: dial tcp 127.0.0.1:7003: connection refused
16:50:30 etcd1 | 2015/03/01 16:50:30 rafthttp: roundtripping error: error roundtripping to http://localhost:7003/raft/stream/a8266ecf031671f3: dial tcp 127.0.0.1:7003: connection refused
16:50:30 etcd2 | 2015/03/01 16:50:30 rafthttp: roundtripping error: error roundtripping to http://localhost:7003/raft/stream/message/6e3bd23ae5f1eae0: dial tcp 127.0.0.1:7003: connection refused
16:50:30 etcd2 | 2015/03/01 16:50:30 rafthttp: roundtripping error: error roundtripping to http://localhost:7003/raft/stream/6e3bd23ae5f1eae0: dial tcp 127.0.0.1:7003: connection refused
16:50:30 etcd1 | 2015/03/01 16:50:30 rafthttp: roundtripping error: error roundtripping to http://localhost:7003/raft/stream/message/a8266ecf031671f3: dial tcp 127.0.0.1:7003: connection refused
16:50:30 etcd1 | 2015/03/01 16:50:30 rafthttp: roundtripping error: error roundtripping to http://localhost:7003/raft/stream/a8266ecf031671f3: dial tcp 127.0.0.1:7003: connection refused
16:50:30 etcd2 | 2015/03/01 16:50:30 rafthttp: roundtripping error: error roundtripping to http://localhost:7003/raft/stream/6e3bd23ae5f1eae0: dial tcp 127.0.0.1:7003: connection refused
16:50:30 etcd2 | 2015/03/01 16:50:30 rafthttp: roundtripping error: error roundtripping to http://localhost:7003/raft/stream/message/6e3bd23ae5f1eae0: dial tcp 127.0.0.1:7003: connection refused
16:50:30 etcd1 | 2015/03/01 16:50:30 rafthttp: roundtripping error: error roundtripping to http://localhost:7003/raft/stream/a8266ecf031671f3: dial tcp 127.0.0.1:7003: connection refused
16:50:30 etcd1 | 2015/03/01 16:50:30 rafthttp: roundtripping error: error roundtripping to http://localhost:7003/raft/stream/message/a8266ecf031671f3: dial tcp 127.0.0.1:7003: connection refused
16:50:30 etcd2 | 2015/03/01 16:50:30 rafthttp: roundtripping error: error roundtripping to http://localhost:7003/raft/stream/6e3bd23ae5f1eae0: dial tcp 127.0.0.1:7003: connection refused
16:50:30 etcd2 | 2015/03/01 16:50:30 rafthttp: roundtripping error: error roundtripping to http://localhost:7003/raft/stream/message/6e3bd23ae5f1eae0: dial tcp 127.0.0.1:7003: connection refused
16:50:31 etcd1 | 2015/03/01 16:50:31 rafthttp: roundtripping error: error roundtripping to http://localhost:7003/raft/stream/a8266ecf031671f3: dial tcp 127.0.0.1:7003: connection refused
16:50:31 etcd1 | 2015/03/01 16:50:31 rafthttp: roundtripping error: error roundtripping to http://localhost:7003/raft/stream/message/a8266ecf031671f3: dial tcp 127.0.0.1:7003: connection refused
16:50:31 etcd2 | 2015/03/01 16:50:31 rafthttp: roundtripping error: error roundtripping to http://localhost:7003/raft/stream/6e3bd23ae5f1eae0: dial tcp 127.0.0.1:7003: connection refused
16:50:31 etcd2 | 2015/03/01 16:50:31 rafthttp: roundtripping error: error roundtripping to http://localhost:7003/raft/stream/message/6e3bd23ae5f1eae0: dial tcp 127.0.0.1:7003: connection refused
16:50:31 etcd1 | 2015/03/01 16:50:31 rafthttp: roundtripping error: error roundtripping to http://localhost:7003/raft/stream/a8266ecf031671f3: dial tcp 127.0.0.1:7003: connection refused
16:50:31 etcd1 | 2015/03/01 16:50:31 rafthttp: roundtripping error: error roundtripping to http://localhost:7003/raft/stream/message/a8266ecf031671f3: dial tcp 127.0.0.1:7003: connection refused
*16:50:31 etcd2 | 2015/03/01 16:50:31 raft: 6e3bd23ae5f1eae0 received msgHeartbeatResp from 924e2e83e93f2560 and changed it to be reachable [next = 15232, match = 0, wait = 0]*
16:50:31 etcd2 | 2015/03/01 16:50:31 raft: 6e3bd23ae5f1eae0 received msgApp rejection(lastindex: 15236) from 924e2e83e93f2560 for index 15237
16:50:31 etcd2 | 2015/03/01 16:50:31 raft: 6e3bd23ae5f1eae0 decreased progress of 924e2e83e93f2560 to [next = 15237, match = 15236, wait = 0]

File: rafthttp/pipeline_test.go [hunk 1]
```diff
@@ -50,7 +50,7 @@ func TestPipelineSend(t *testing.T) {
 func TestPipelineExceedMaximalServing(t *testing.T) {
 	tr := newRoundTripperBlocker()
 	fs := &stats.FollowerStats{}
-	p := newPipeline(tr, "http://10.0.0.1", types.ID(1), types.ID(1), fs, nil)
+	p := newPipeline(tr, "http://10.0.0.1", types.ID(1), types.ID(1), fs, &nopProcessor{}, nil)
```
Comment (xiang90) [inline rafthttp/pipeline_test.go]: i still do not think it should be a processor as i have mentioned in other pull requests. but i think we need to address this in a separate commit.

**Comment by xiang90** [comment]
@yichengq The reader side can also report i guess. But this pr LGTM.

File: rafthttp/pipeline_test.go (hunk 1 above)
Comment (yichengq) [inline rafthttp/pipeline_test.go]: yup

**Pull request merged**